doc/psidk/src/concepts/disk-storage.md

```diff
@@ -0,0 +1,163 @@
+
+Successfully peering a verified file from a third-party can be thought of as a proof of work. Therefore, a Psibase network could even consider a standard network token issuance directed to those who service peer download requests from IPs proportionally to the bandwidth they service. This would be a way for a network to subsidize greater file access decentralization.
+
+## Infrastructure providers
+
+### Running an infrastructure provider
+
+To run an infrastructure provider, one needs to run a full node that also contains all pinned IPFS files. Registering as an infrastructure provider is intended to be an implicit claim that one's node contains all pinned IPFS data. If this is true for the majority of infrastructure providers and candidates, then even large changes to the active IP set should be stable and continually allow a majority of IPs to offer download services for the pinned IPFS files. This rule cannot be enforced objectively, because it is not possible to prove that another node has downloaded all files without individually downloading and validating each file from them, which would consume excessive bandwidth. However, a node that does not have all CIDs downloaded but who is elected to be an active Infrastructure Provider will either be unable to serve various resources, or may proxy an excessive number of web requests to other infrastructure providers. Both behaviors are therefore detectable and subjective community consensus can either remove them, slash their bond, or both.
+
+When starting up a new infrastructure provider, the node will peer all IPFS data from the IPFS network. At minimum, therefore, the user running the infrastructure provider node must be able to pay for the bandwidth necessary to transfer all pinned files.
+
+### Publishing upload rates
+
+Every IP must publish their current rate that they charge for upload bandwidth. If they fail to set a rate, users will be able to consume the infrastructure provider's network bandwidth for free. This rate is then used to enforce a maximum billable amount after an upload is performed. Furthermore, as infrastructure providers adjust their rates, browser clients can automatically select the cheapest upload service providers to effectively achieve decentralized load balancing. <span style="color:red"> TODO - However, don't we have the same issue I described for download? An IP could set an excessively large upload rate to effectively outsource their upload responsibility to others, centralizing the service provision. If this was a shared fee instead, it would also simplify calculating the bond amount for third-party IPFS clients.</span>
```

Comment:
> An IP could set an excessively large upload rate to effectively outsource their upload responsibility to others, centralizing the service provision. If this was a shared fee instead, it would also simplify calculating the bond amount for third-party IPFS clients.

This still seems to me to imply that we must use a shared rate for the upload bandwidth cost to avoid IPs implicitly opting out of providing this service.